File: Server/csharp/AppToken/Model/Service.cs
using System.Text;
using AppToken.Util;

namespace AppToken.Model
{
    public class Service
    {
        private readonly string channelId;
        private readonly string userId;
        private int? privilege;

        public Service(string channelId, string userId)
        {
            this.channelId = channelId ?? throw new ArgumentNullException(nameof(channelId));
            this.userId = userId ?? throw new ArgumentNullException(nameof(userId));
        }

        private Service(string channelId, string userId, int? privilege)
        : this(channelId, userId)
        {
            this.privilege = privilege;
        }

        public static Service CreateServiceOnlyWithUserId(string userId)
        {
            return new Service("*", userId);
        }

        public static Service CreateServiceOnlyWithChannelId(string channelId)
        {
            return new Service(channelId, "*");
        }

        public void AddAudioPublishPrivilege()
        {
            if (!this.privilege.HasValue)
            {
                this.privilege = 0;
                this.privilege |= PrivilegeEnum.ENABLE_PRIVILEGE;
            }
            this.privilege |= PrivilegeEnum.ENABLE_AUDIO_PUBLISH;
        }

        public void AddVideoPublishPrivilege()
        {
            if (!this.privilege.HasValue)
            {
                this.privilege = 0;
                this.privilege |= PrivilegeEnum.ENABLE_PRIVILEGE;
            }
            this.privilege |= PrivilegeEnum.ENABLE_VIDEO_PUBLISH;
        }

        public void AddScreenPublishPrivilege()
        {
            if (!this.privilege.HasValue)
            {
                this.privilege = 0;
                this.privilege |= PrivilegeEnum.ENABLE_PRIVILEGE;
            }
            this.privilege |= PrivilegeEnum.ENABLE_SCREEN_PUBLISH;
        }

        public void Pack(MemoryStream buf)
        {
            using (var writer = new BinaryWriter(buf, Encoding.UTF8, true))
            {
                var channelIdBytes = Encoding.UTF8.GetBytes(this.channelId);
                writer.Write(BigEndianUtils.GetBytesBigEndian(channelIdBytes.Length));
                writer.Write(channelIdBytes);

                var userIdBytes = Encoding.UTF8.GetBytes(this.userId);
                writer.Write(BigEndianUtils.GetBytesBigEndian(userIdBytes.Length));
                writer.Write(userIdBytes);

                writer.Write(this.privilege.HasValue);

                if (this.privilege.HasValue)
                {
                    writer.Write(BigEndianUtils.GetBytesBigEndian(this.privilege.Value));
                }
            }
        }

        public static Service Unpack(MemoryStream buf)
        {
            using (var reader = new BinaryReader(buf, Encoding.UTF8, true))
            {
                var channelLength = BigEndianUtils.FromBytesBigEndianToInt(reader.ReadBytes(4));
                var byteChannelId = reader.ReadBytes(channelLength);
                var channelId = Encoding.UTF8.GetString(byteChannelId);

                var userIdLength = BigEndianUtils.FromBytesBigEndianToInt(reader.ReadBytes(4));
                var byteUserId = reader.ReadBytes(userIdLength);
                var userId = Encoding.UTF8.GetString(byteUserId);

                var hasPrivilege = reader.ReadBoolean();
                int? privilege = hasPrivilege ? BigEndianUtils.FromBytesBigEndianToInt(reader.ReadBytes(4)) : (int?)null;

                return new Service(channelId, userId, privilege);
            }
        }
    }

    public static class PrivilegeEnum
    {
        public const int ENABLE_PRIVILEGE = 1;
        public const int ENABLE_AUDIO_PUBLISH = 2;
        public const int ENABLE_VIDEO_PUBLISH = 4;
        public const int ENABLE_SCREEN_PUBLISH = 8;
    }
}

File: Android/DingRTC-APIExample/common/token_generator/src/main/java/com/ding/rtc/api/example/token_generator/model/Service.java
package com.ding.rtc.api.example.token_generator.model;

import com.ding.rtc.api.example.token_generator.constant.AppTokenConstants;
import com.ding.rtc.api.example.token_generator.enums.PrivilegeEnum;
import io.netty.buffer.ByteBuf;
import lombok.Getter;
import lombok.NonNull;

import java.util.Objects;

@Getter
public class Service {
    @NonNull
    private final String channelId;
    @NonNull
    private final String userId;
    private Integer privilege;

    public Service(@NonNull String channelId, @NonNull String userId) {
        this.channelId = channelId;
        this.userId = userId;
    }

    public static Service createServiceOnlyWithUserId(@NonNull String userId) {
        return new Service(AppTokenConstants.WILDCARD_CHARACTERS, userId);
    }

    public static Service createServiceOnlyWithChannelId(@NonNull String channelId) {
        return new Service(channelId, AppTokenConstants.WILDCARD_CHARACTERS);
    }

    public void addAudioPublishPrivilege() {
        if (Objects.isNull(this.privilege)) {
            this.privilege = 0;
            this.privilege = this.privilege | PrivilegeEnum.ENABLE_PRIVILEGE.getPrivilege();
        }
        this.privilege = this.privilege | PrivilegeEnum.ENABLE_AUDIO_PUBLISH.getPrivilege();
    }

    public void addVideoPublishPrivilege() {
        if (Objects.isNull(this.privilege)) {
            this.privilege = 0;
            this.privilege = this.privilege | PrivilegeEnum.ENABLE_PRIVILEGE.getPrivilege();
        }
        this.privilege = this.privilege | PrivilegeEnum.ENABLE_VIDEO_PUBLISH.getPrivilege();
    }

    public void addScreenPublishPrivilege() {
        if (Objects.isNull(this.privilege)) {
            this.privilege = 0;
            this.privilege = this.privilege | PrivilegeEnum.ENABLE_PRIVILEGE.getPrivilege();
        }
        this.privilege = this.privilege | PrivilegeEnum.ENABLE_SCREEN_PUBLISH.getPrivilege();
    }


    public void pack(@NonNull ByteBuf buf) {
        final byte[] channelId = this.channelId.getBytes();
        buf.writeInt(channelId.length);
        buf.writeBytes(channelId);
        final byte[] userId = this.userId.getBytes();
        buf.writeInt(userId.length);
        buf.writeBytes(userId);
        final boolean hasPrivilege = Objects.nonNull(this.privilege);
        buf.writeBoolean(hasPrivilege);
        if (hasPrivilege) {
            buf.writeInt(this.privilege);
        }
    }

    public static Service unpack(@NonNull ByteBuf buf) {
        final byte[] byteChannelId = new byte[buf.readInt()];
        for (int i = 0; i < byteChannelId.length; i++) {
            byteChannelId[i] = buf.readByte();
        }
        final String channelId = new String(byteChannelId);

        final byte[] byteUserId = new byte[buf.readInt()];
        for (int i = 0; i < byteUserId.length; i++) {
            byteUserId[i] = buf.readByte();
        }
        final String userId = new String(byteUserId);

        final boolean hasPrivilege = buf.readBoolean();
        Integer privilege = null;
        if (hasPrivilege) {
            privilege = buf.readInt();
        }
        return new Service(channelId, userId, privilege);
    }

    private Service(@NonNull String channelId, @NonNull String userId, Integer privilege) {
        this.channelId = channelId;
        this.userId = userId;
        this.privilege = privilege;
    }
}
